### ISSUE TITLE
Drop Connect Layer

### ISSUE THREAD

**Issue by mr-raccoon-97** [pytorch/pytorch#131688]
### 🚀 The feature, motivation and pitch

I was reading the Yann LeCun's Drop Connect paper:
[](https://proceedings.mlr.press/v28/wan13.html)
And didn't find any implementation on the internet.

### Alternatives

So I created my own, here: 
[](https://github.com/mr-raccoon-97/drop-connect/tree/main)

The implementation of DropConnect layers is here:
[](https://github.com/mr-raccoon-97/drop-connect/blob/main/model/dropconnect.py)

But also made an alternative version that disables the weight drops during the evaluation phase, like the dropout.
[](https://github.com/mr-raccoon-97/drop-connect/blob/main/model/batched_dropweight.py)

### Additional context

The are no papers avaliable of the usage of drop connect layers, because there isn't any implementation avaliable on the internet. There are some implementation of weight drop saying they are drop connect, but the don't implement the layers as stated in the paper. 

My solution works, and works well.

**Comment by janeyx99** [comment]
Hi, thanks for bringing this to our attention. Do you have a particular ask for PyTorch? Right now, it looks like having it live in a separate repo is perfectly acceptable.

**Comment by mr-raccoon-97** [comment]
> Hi, thanks for bringing this to our attention. Do you have a particular ask for PyTorch? Right now, it looks like having it live in a separate repo is perfectly acceptable.

I'm not expecting you to copy and paste my solution in PyTorch, but it would be nice to have some official implementation of DropConnect layers, since it's error prone to just hack one, and see many people implementing it in another way or just plain wrong.

For example:
[https://discuss.pytorch.org/t/dropconnect-implementation/70921/7](https://discuss.pytorch.org/t/dropconnect-implementation/70921/7) 

[https://discuss.pytorch.org/t/is-the-pytorch-nlp-dropconnect-implementation-correct/96282/2]
(https://discuss.pytorch.org/t/is-the-pytorch-nlp-dropconnect-implementation-correct/96282/2)

Or the implementation here, it's just different from the paper.
[https://paperswithcode.com/method/dropconnect](https://paperswithcode.com/method/dropconnect)

[PyTorch NLP library](https://pytorchnlp.readthedocs.io/en/latest/_modules/torchnlp/nn/weight_drop.html) has one, but dropconnect is not something from nlp, also it's not implemented as a trainable layer.

**Comment by janeyx99** [comment]
Thanks for the details!

**Comment by geleshChrsitUniversity** [comment]
One of the primary reasons DropConnect is absent from native frameworks like PyTorch, TensorFlow, or Keras is that the **original work by Wan et al. (which included Yann LeCun)** predates these modern frameworks and required a completely different engineering architecture.The original paper highlighted massive memory and bookkeeping challenges when masking individual weights within a matrix.

To overcome this, the authors had to write custom **CUDA kernels utilizing highly specialized engineering like bit-packing and 2D texture-aligned memory**. It essentially required rebuilding the underlying GPU engine—a feat they accomplished by using their own customized, non-generic framework.In contrast, modern frameworks like PyTorch and Keras were built to act as generic computational graphs. _**Frameworks such as pytorch are optimized to host standardized, modular layers (like Dense or MaxPool) rather than altering underlying tensor connections on the fly.**_

As a result, community workarounds I have seen so far, simply apply one weight mask shared across the entire batch. **True DropConnect, however, strictly requires a unique weight mask per individual sample/row within the batch**. Implementing this true, sample-wise behavior is incredibly inefficient and challenging to do "as is" within the standard APIs of PyTorch or Keras without suffering a catastrophic drop in training speed.
